strawberryfields/backends/states.py

```diff
@@ -832,6 +833,74 @@ def expand_dims(op, n, modes):
 
         return mean, var
 
+    def number_expectation(self, modes):
+        r"""
+        Calculates the expectation value of the product of the number operators of the modes.
+
+        This method computes the analytic expectation value
+        :math:`\langle \hat{n}_{i_0} \hat{n}_{i_1}\dots \hat{n}_{i_{N-1}}\rangle`
+        for a (sub)set of modes :math:`[i_0, i_1, \dots, i_{N-1}]` of the system.
+
```

Comment:
Now I got myself wondering about the advantages of having that class which is "empty" :blush:  It might make sense to create this facade-like abstract method in `BaseState` and include docstring containing the formula there. The reason is, that in the documentation it would appear for both the `BaseFockState` and for the `BaseGaussianState`. The example that is included could stay specific to either, we could just have a comment that users should be replaced in the other case. This would also just help us to make sure that we have only one docstring that we need to maintain overall. Let me know if I can help with the details on this.

Reply:
Yup, maybe you can implement this. Is it appropriate to have it in this PR?

Reply:
Sure, I think so! I'll get around to it :)